### PR TITLE
`Development`: Fix code coverage script not found files marked as deleted

### LINKED ISSUE #7035: Coverage script falsely displays `deleted` in line coverage column
### Describe the bug

Currently, our support script to generate the coverage table falsely reports `deleted` when the coverage could not be retrieved.

### To Reproduce

Generate a code coverage table for a branch that contains a file for which no coverage can be retrieved.

### Expected behavior

If the coverage can't be retrieved it should display a dedicated flag (e.g. `not found`)

### Screenshots

Line of coverage table:
Screenshot 2023-08-03 at 13 30 30

Proof that file was not deleted:
Screenshot 2023-08-03 at 13 31 07

### Which version of Artemis are you seeing the problem on?

6.3.9

### What browsers are you seeing the problem on?

Other (specify in "Additional context")

### Additional context

Not related to browser.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by FelixTJDietrich**
#### General

- [x] This is a small issue that I tested locally.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context

Closes #7035

### Description

Changed `deleted` to `not found (<change_type>)` if the line coverage was not found. It uses the git index change type of the non-test file which is either `added`, `deleted`, `modified`, `renamed`, or `unknown`.

Additionally, there are some minor improvements and a check for the existence of the test artefact with a corresponding message, rather than just writing `not found` for each file.

### Steps for Testing

1. `cd supporting_scripts/generate_code_cov_table`
2. Follow the `Readme.md` if you have not used the script before
3. Run `python generate_code_cov_table.py --branch-name feature/learning-path/feature-toggle` (or any other branch that has a PR where coverage results exist)
4. Check that `deleted` is now `not found (<change_type>)`

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

**Example: https://github.com/ls1intum/Artemis/pull/7025**

If a line coverage has not been found it is now `not found (<change_type>)` as seen below.

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [user.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/user/user.model.ts.html) | 100% | ✅ |
| [competency-management.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/competencies/competency-management/competency-management.component.ts.html) | 79.78% | ✅❌ |
| [learning-path-management.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/learning-path-management/learning-path-management.component.ts.html) | 100% | ✅ |
| [learning-path-paging.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/learning-path-paging.service.ts.html) | 90.9% | ✅❌ |
| [learning-path.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/learning-paths/learning-path.service.ts.html) | 100% | ✅ |
| [course-management-tab-bar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-management-tab-bar/course-management-tab-bar.component.ts.html) | 96.77% | ✅❌ |
| course-management.route.ts | not found (modified) | ❌ |
| [learning-path-health.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/competency/learning-path-health.model.ts.html) | 100% | ✅ |
| [learning-path.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/competency/learning-path.model.ts.html) | 100% | ✅ |
| [course.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/course.model.ts.html) | 100% | ✅ |
| [feature-toggle.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/feature-toggle/feature-toggle.service.ts.html) | 65.71% | ✅❌ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 89.89% | ✅❌ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [Course.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/Course.html) | 91% | ✅❌ |
| [User.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/User.html) | 90% | ✅❌ |
| [Competency.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/Competency.html) | 79% | ✅❌ |
| [LearningPath.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/LearningPath.html) | 87% | ✅❌ |
| CompetencyProgressRepository.java | not found (modified) | ❌ |
| [CourseRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/CourseRepository.html) | 77% | ✅❌ |
| [LearningPathRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/LearningPathRepository.html) | 61% | ✅❌ |
| [LectureUnitRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/LectureUnitRepository.html) | 36% | ✅❌ |
| [UserRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/UserRepository.html) | 92% | ✅❌ |
| [CompetencyProgressService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/CompetencyProgressService.html) | 91% | ✅❌ |
| [CourseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/CourseService.html) | 89% | ✅❌ |
| [LearningPathService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/LearningPathService.html) | 98% | ✅❌ |
| [Feature.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.feature/Feature.html) | 100% | ✅ |
| [FeatureToggleService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.feature/FeatureToggleService.html) | 100% | ✅ |
| [CompetencyResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CompetencyResource.html) | 97% | ✅❌ |
| [CourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CourseResource.html) | 94% | ✅❌ |
| [LearningPathResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/LearningPathResource.html) | 94% | ✅❌ |
| [LearningPathHealthDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.competency/LearningPathHealthDTO.html) | 100% | ✅ |
| [LearningPathPageableSearchDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.competency/LearningPathPageableSearchDTO.html) | 100% | ✅ |
| [UserNameAndLoginDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.user/UserNameAndLoginDTO.html) | 100% | ✅ |
| [PageUtil.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5270/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.util/PageUtil.html) | 93% | ✅❌ |

### CODE REVIEW

**Review by MaximilianAnzinger — APPROVED** [review]
LGTM 👍 Thank you for adding this detail.

**Comment by MaximilianAnzinger** [comment]
Just one question:
You check the changed files for the creation of this table. Wouldn't this also allow you to immediately check if the file was edited/deleted? If I understand the GitPython library correctly, the `DiffIndex`s in `get_changed_files` should include the change type.

**Comment by FelixTJDietrich** [comment]
> Just one question:
> You check the changed files for the creation of this table. Wouldn't this also allow you to immediately check if the file was edited/deleted? If I understand the GitPython library correctly, the `DiffIndex`s in `get_changed_files` should include the change type.

Yeah I was wondering how you would handle renames for example, I will look into it 🤔 I tried to do it manually, but maybe I can just do as you said.

**Comment by MaximilianAnzinger** [comment]
That's another edge case I didn't think of 👍 
But this should also be easily detectable.

> change_type –
Member of DiffIndex.change_type, namely:
> - ’A’ for added paths
> - ’D’ for deleted paths
> - ’R’ for renamed paths
> - ’M’ for paths with modified data
> - ’T’ for changed in the type paths

**Review by pal03377 — APPROVED** [review]
Changes look good 👍 I also tried it out with the GitHub preview to ensure that the new slash somehow doesn't break the markdown table.

Edit: Handling renames in some way would be fantastic as well, of course.

**Comment by FelixTJDietrich** [comment]
Thanks for the tip @MaximilianAnzinger I think it worked :)

**Review by MaximilianAnzinger — APPROVED** [review]
Code changes look good, works very well, and the auto-checking 100% coverage is a great QoL addition 👍

**Review by pal03377 — CHANGES_REQUESTED** [review]
#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [athena.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/assessment/athena.service.ts.html) | 100% | ✅ |
| text-exercise-cluster-statistics.model.ts | not found (added) | ❌ |
| exam-management.route.ts | not found (modified) | ❌ |
...

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [RestTemplateConfiguration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.config/RestTemplateConfiguration.html) | 95% | ✅❌ |
...
| [AthenaSubmissionSendingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaSubmissionSendingService.html) | 91% | ✅❌ |
| AtheneConnector.java | not found (added) | ❌ |
| AtheneHealthIndicator.java | not found (added) | ❌ |
| AtheneService.java | not found (added) | ❌ |
...
| AdminTextExerciseResource.java | not found (added) | ❌ |
| TextClusterStatisticsDTO.java | not found (added) | ❌ |
| AtheneResource.java | not found (added) | ❌ |

---

This is the output for the `development/new-athena` branch. However, I clearly didn't add most of these files but deleted them 🤔

**Comment by MaximilianAnzinger** [comment]
@FelixTJDietrich maybe it's related to [this issue](https://github.com/gitpython-developers/GitPython/issues/882)?

**Comment by FelixTJDietrich** [comment]
@MaximilianAnzinger Thanks for pointing me to the solution! I think it is fixed now.

@pal03377 does this seem right to you?

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [athena.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/assessment/athena.service.ts.html) | 100% | ✅ |
| text-exercise-cluster-statistics.model.ts | not found (deleted) | ❌ |
| exam-management.route.ts | not found (modified) | ❌ |
| [text-submission-assessment.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/assess/text-submission-assessment.component.ts.html) | 92.03% | ✅❌ |
| cluster-statistics.component.ts | not found (deleted) | ❌ |
| cluster-statistics.route.ts | not found (deleted) | ❌ |
| [text-exercise.route.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise.route.ts.html) | 0% | ✅❌ |
| [text-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise.service.ts.html) | 100% | ✅ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 89.89% | ✅❌ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [RestTemplateConfiguration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.config/RestTemplateConfiguration.html) | 95% | ✅❌ |
| AtheneAuthorizationInterceptor.java | not found (renamed) | ❌ |
| [Feedback.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/Feedback.html) | 97% | ✅❌ |
| [TextBlock.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/TextBlock.html) | 96% | ✅❌ |
| [TextBlockRef.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/TextBlockRef.html) | 100% | ✅ |
| TextCluster.java | not found (deleted) | ❌ |
| [TextExercise.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/TextExercise.html) | 93% | ✅❌ |
| NetworkingError.java | not found (modified) | ❌ |
| [FeedbackRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/FeedbackRepository.html) | 83% | ✅❌ |
| TextBlockRepository.java | not found (modified) | ❌ |
| TextClusterRepository.java | not found (deleted) | ❌ |
| [TextSubmissionRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/TextSubmissionRepository.html) | 88% | ✅❌ |
| AutomaticTextFeedbackService.java | not found (deleted) | ❌ |
| [ExerciseDeletionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ExerciseDeletionService.html) | 96% | ✅❌ |
| [SubmissionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/SubmissionService.html) | 91% | ✅❌ |
| TextAssessmentQueueService.java | not found (deleted) | ❌ |
| [TextAssessmentService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/TextAssessmentService.html) | 100% | ✅ |
| [TextExerciseImportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/TextExerciseImportService.html) | 100% | ✅ |
| [TextSubmissionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/TextSubmissionService.html) | 88% | ✅❌ |
| [ApollonHealthIndicator.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.apollon/ApollonHealthIndicator.html) | 100% | ✅ |
| AtheneConnector.java | not found (renamed) | ❌ |
| [AthenaFeedbackSendingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaFeedbackSendingService.html) | 93% | ✅❌ |
| [AthenaFeedbackSuggestionsService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaFeedbackSuggestionsService.html) | 83% | ✅❌ |
| [AthenaHealthIndicator.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaHealthIndicator.html) | 95% | ✅❌ |
| [AthenaSubmissionSelectionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaSubmissionSelectionService.html) | 100% | ✅ |
| [AthenaSubmissionSendingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.athena/AthenaSubmissionSendingService.html) | 91% | ✅❌ |
| AtheneHealthIndicator.java | not found (deleted) | ❌ |
| AtheneService.java | not found (deleted) | ❌ |
| [TextExerciseDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.dto.athena/TextExerciseDTO.html) | 100% | ✅ |
| [TextFeedbackDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.dto.athena/TextFeedbackDTO.html) | 91% | ✅❌ |
| [TextSubmissionDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.dto.athena/TextSubmissionDTO.html) | 100% | ✅ |
| [DistributedInstanceMessageSendService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/DistributedInstanceMessageSendService.html) | 0% | ✅❌ |
| [InstanceMessageReceiveService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/InstanceMessageReceiveService.html) | 74% | ✅❌ |
| InstanceMessageSendService.java | not found (modified) | ❌ |
| [MainInstanceMessageSendService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/MainInstanceMessageSendService.html) | 95% | ✅❌ |
| [MessageTopic.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/MessageTopic.html) | 100% | ✅ |
| [AthenaScheduleService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.scheduled/AthenaScheduleService.html) | 44% | ✅❌ |
| AtheneScheduleService.java | not found (deleted) | ❌ |
| [AthenaResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/AthenaResource.html) | 73% | ✅❌ |
| [TextAssessmentResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/TextAssessmentResource.html) | 90% | ✅❌ |
| TextClusterResource.java | not found (deleted) | ❌ |
| [TextSubmissionResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/TextSubmissionResource.html) | 97% | ✅❌ |
| AdminTextExerciseResource.java | not found (deleted) | ❌ |
| TextClusterStatisticsDTO.java | not found (deleted) | ❌ |
| AtheneResource.java | not found (deleted) | ❌ |

**Comment by MaximilianAnzinger** [comment]
@FelixTJDietrich Thx for fixing this 👍 
For now, this looks great 🎉 . 

I've a follow-up question? Is the script to work for PR's that don't merge into `develop`? If so I might have encountered another bug.

**Comment by FelixTJDietrich** [comment]
@MaximilianAnzinger I think there is a CLI option where you can set the base branch try --help, if this is what you need 🤔

**Comment by MaximilianAnzinger** [comment]
@FelixTJDietrich for this PR: https://github.com/ls1intum/Artemis/pull/7061
The script detects 3 changed files, but it should be 11.

Output:
```
generate_code_cov_table % python generate_code_cov_table.py --username <username> --password <password> --base-branch-name feature/learning-path/feature-toggle
Using current branch: bugfix/learning-path/course-edit
Using latest build ID: 5303
Fetching client coverage: 100%|████████████████| 2/2 [00:00<00:00,  3.26files/s]
Fetching server coverage: 100%|████████████████| 1/1 [00:00<00:00,  3.81files/s]
Info: ✅❌ in Confirmation (assert/expect) have to be adjusted manually, also delete trivial files!

Code coverage report copied to clipboard. Use --print-results to print it to console instead.
```

**Review by MaximilianAnzinger — DISMISSED** [review]

**Review by reschandreas — COMMENTED** [review]
Code LGTM, I have only two small suggestions that would fix the output in my shell. Do the emojis not overlap in your terminal? If not, then ignore my suggestions :)

**Comment by FelixTJDietrich** [comment]
@reschandreas Thanks for reviewing and the suggestion! I guess this is somewhat unrelated to this PR but I just added it.

**Review by pal03377 — COMMENTED** [review]
The script now doesn't find almost any of the files, claiming for all of them to be "modified" 😕  I'm pretty sure most of them should be linkable.

Example for Athena Integration:

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [athena.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/assessment/athena.service.ts.html) | 100% | ✅ |
| text-exercise-cluster-statistics.model.ts | not found (deleted) | ❌ |
| exam-management.route.ts | not found (modified) | ❌ |
| [text-submission-assessment.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/assess/text-submission-assessment.component.ts.html) | 92.03% | ✅ ❌ |
| cluster-statistics.component.ts | not found (deleted) | ❌ |
| cluster-statistics.route.ts | not found (deleted) | ❌ |
| [text-exercise.route.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise.route.ts.html) | 0% | ✅ ❌ |
| [text-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise.service.ts.html) | 100% | ✅ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 89.89% | ✅ ❌ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| RestTemplateConfiguration.java | not found (modified) | ❌ |
| AtheneAuthorizationInterceptor.java | not found (renamed) | ❌ |
| Feedback.java | not found (modified) | ❌ |
| TextBlock.java | not found (modified) | ❌ |
| TextBlockRef.java | not found (added) | ❌ |
| TextCluster.java | not found (deleted) | ❌ |
| TextExercise.java | not found (modified) | ❌ |
| NetworkingError.java | not found (deleted) | ❌ |
| NetworkingException.java | not found (added) | ❌ |
| FeedbackRepository.java | not found (modified) | ❌ |
| GradingInstructionRepository.java | not found (modified) | ❌ |
| TextBlockRepository.java | not found (modified) | ❌ |
| TextClusterRepository.java | not found (deleted) | ❌ |
| TextSubmissionRepository.java | not found (modified) | ❌ |
| AutomaticTextFeedbackService.java | not found (deleted) | ❌ |
| ExerciseDeletionService.java | not found (modified) | ❌ |
| SubmissionService.java | not found (modified) | ❌ |
| TextAssessmentQueueService.java | not found (deleted) | ❌ |
| TextAssessmentService.java | not found (modified) | ❌ |
| TextExerciseImportService.java | not found (modified) | ❌ |
| TextSubmissionService.java | not found (modified) | ❌ |
| ApollonHealthIndicator.java | not found (modified) | ❌ |
| AthenaConnector.java | not found (added) | ❌ |
| AthenaFeedbackSendingService.java | not found (added) | ❌ |
| AthenaFeedbackSuggestionsService.java | not found (added) | ❌ |
| AthenaHealthIndicator.java | not found (added) | ❌ |
| AthenaSubmissionSelectionService.java | not found (added) | ❌ |
| AthenaSubmissionSendingService.java | not found (added) | ❌ |
| AtheneConnector.java | not found (deleted) | ❌ |
| AtheneHealthIndicator.java | not found (deleted) | ❌ |
| AtheneService.java | not found (deleted) | ❌ |
| TextExerciseDTO.java | not found (added) | ❌ |
| TextFeedbackDTO.java | not found (added) | ❌ |
| TextSubmissionDTO.java | not found (added) | ❌ |
| DistributedInstanceMessageSendService.java | not found (modified) | ❌ |
| InstanceMessageReceiveService.java | not found (modified) | ❌ |
| InstanceMessageSendService.java | not found (modified) | ❌ |
| MainInstanceMessageSendService.java | not found (modified) | ❌ |
| MessageTopic.java | not found (modified) | ❌ |
| AthenaScheduleService.java | not found (added) | ❌ |
| AtheneScheduleService.java | not found (deleted) | ❌ |
| AthenaResource.java | not found (added) | ❌ |
| TextAssessmentResource.java | not found (modified) | ❌ |
| TextClusterResource.java | not found (deleted) | ❌ |
| TextSubmissionResource.java | not found (modified) | ❌ |
| AdminTextExerciseResource.java | not found (deleted) | ❌ |
| TextClusterStatisticsDTO.java | not found (deleted) | ❌ |
| AtheneResource.java | not found (deleted) | ❌ |

File: supporting_scripts/generate_code_cov_table/generate_code_cov_table.py [hunk 1]
```diff
@@ -73,30 +73,40 @@ def get_changed_files(branch_name, base_branch_name="origin/develop"):
...
-    server_file_names = []
 
-    for file_name in file_names:
+def filter_file_changes(file_changes):
+    client_file_changes = {}
+    server_file_changes = {}
+
+    for file_name, change_type in file_changes.items():
         if file_name.startswith("src/main/webapp/app"):
             if file_name.endswith(".ts") and "module.ts" not in file_name:
```
Comment (pal03377) [inline supporting_scripts/generate_code_cov_table/generate_code_cov_table.py:98]: Not really the subject of this PR, but I just noticed this condition and wondered why the `module.ts` part is needed 🤔 Also, could it be this then?
```suggestion
            if file_name.endswith(".ts") and not file_name.endswith("module.ts"):
```

Comment (FelixTJDietrich) [reply supporting_scripts/generate_code_cov_table/generate_code_cov_table.py:98]: I guess we can add this ^^, I'm adding some other check as well. This is more reliable. The module files never seem to have an associated test file as far as I could see, that is why I ignore them.

**Comment by FelixTJDietrich** [comment]
Thank's @pal03377 for finding this. I'm now checking if the artefacts exist and output a message to the table to make it less confusing. 

**Example:**

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [athena.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/assessment/athena.service.ts.html) | 100% | ✅ |
| text-exercise-cluster-statistics.model.ts | not found (deleted) | ❌ |
| exam-management.route.ts | not found (modified) | ❌ |
| [text-submission-assessment.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/assess/text-submission-assessment.component.ts.html) | 92.03% | ✅ ❌ |
| cluster-statistics.component.ts | not found (deleted) | ❌ |
| cluster-statistics.route.ts | not found (deleted) | ❌ |
| [text-exercise.route.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise.route.ts.html) | 0% | ✅ ❌ |
| [text-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise.service.ts.html) | 100% | ✅ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5113/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 89.89% | ✅ ❌ |

#### Server

Coverage report not found, tests probably failed.

**Review by dearjasmina — APPROVED** [review]
code changes look good 👍🏻

**Review by reschandreas — APPROVED** [review]
Code LGTM

**Review by bassner — APPROVED** [review]

**Pull request merged**